scripts/zones/Aht_Urhgan_Whitegate/npcs/Kaduru-Haiduru.lua

```diff
@@ -0,0 +1,100 @@
+    local shihuDanhuEncounters = player:getVar("ShihuDanhu_Encounters")
+    local shihuDanhuDate = player:getVar("ShihuDanhu_TP_date")
+    local currentDate = os.date("%j")
+
+    -- Kaduru-Haiduru can be used unless the following are true.
+    if (currentDate - shihuDanhuDate < 1 and shihuDanhuEncounters > 1) or
+        (currentDate - caughtUsingShihuDanhuDate < 1) then
+        return false
+    end
+    return true
+end;
+
+function onTrigger(player, npc)
+    local caughtUsingShihuDanhuDate = player:getVar("Kaduru_ShihuDanhu_date")
+    local shihuDanhuDate = player:getVar("ShihuDanhu_TP_date")
+    local timesUsed = player:getVar("Kaduru_TimesUsed")
+    local currentDate = os.date("%j")
+
+    if canUse_KaduruHaiduru_Service(player) then
+        player:startEvent(151, 0, 0, timesUsed, 0, 0, 0, 0, 0, 0)
+    else
+        if caughtUsingShihuDanhuDate == 0 then
+            player:setVar("Kaduru_ShihuDanhu_date", os.date("%j"))
+            player:setVar("Kaduru_TimesUsed", 0)
+        end
+        player:startEvent(153, 0, 0, 0, 0, 0, 0, 0, 0, 0)
+    end
+end
+
+function onTrade(player, npc, trade)
+    local caughtUsingShihuDanhuDate = player:getVar("Kaduru_ShihuDanhu_date")
+    local shihuDanhuDate = player:getVar("ShihuDanhu_TP_date")
+    local timesUsed = player:getVar("Kaduru_TimesUsed")
+
+    if canUse_KaduruHaiduru_Service(player) and timesUsed == 3 then
+        if trade:getItemCount() == 1 then
+            if trade:hasItemQty(2185, 1) then
+                player:startEvent(154, 0, player:getNation(), 0, 0, 0, 0, 0, 0, 0)
+                player:setVar("ShihuDanhu_TP_date", 0)
+                player:setVar("Kaduru_ShihuDanhu_date", 0)
+            end
+        end
+    elseif canUse_KaduruHaiduru_Service(player) and timesUsed < 3 then
+        if trade:getItemCount() == 1 then
+            if trade:hasItemQty(2185, 1) then
+                player:startEvent(152, 0, 0, 0, 0, 0, 0, 0, 0, 0)
+                player:setVar("ShihuDanhu_TP_date", 0)
+                player:setVar("Kaduru_ShihuDanhu_date", 0)
+            end
+        end
+    else
+        if caughtUsingShihuDanhuDate == 0 then
+            player:setVar("Kaduru_ShihuDanhu_date", os.date("%j"))
+            player:setVar("Kaduru_TimesUsed", 0)
+        end
+        player:startEvent(155, 0, 0, 0, 0, 0, 0, 0, 0, 0)
+    end
+end
+
+function onEventUpdate(player, csid, option) 
+end;
+
+function onEventFinish(player, csid, option)
+    local timesUsed = player:getVar("Kaduru_TimesUsed")
+    if csid == 154 then        -- At this point we should already have used Kaduru-Haiduru 3 times.
+        if option == 1 then       -- Duchy of Jeuno
+            player:setPos(0, 3, -6, 190, 243)
+        elseif option == 2 then   -- Nation of Allegiance
+            local nation = player:getNation()
+            switch (nation): caseof
+            {
+                [0] = function(x)  -- San d'Oria
```

Comment:
can use position tables to avoid the weird lua-implemented switch, or stacked elseifs

```
local nationDestinations =
{
    [dsp.nation.SANDORIA] = {110,   0,  -7, 175, 231},
    [dsp.nation.BASTOK  ] = { 90,   0, -67, 248, 234},
    [dsp.nation.WINDURST] = {192, -12, 218,  60, 240},
}
local destination = nationDestinations[player:getNation()]
player:setPos(unpack(destination))

local randomDestinations =
{
    {  -1.015,  8.999,  -52.962, 192, 243}, -- Ru Lude Gardens (H-9)   {R}
    { 382.398,  7.314, -106.298, 160, 105}, -- Batallia Downs (K-8)    {R}
    {-327.238,  2.914,  438.421, 130, 120}, -- Sauromugue Champaign    {R}
    { 213.785, 16.356,  419.961, 218, 110}, -- Rolanberry Fields (J-6) {R}
    { 167.093, 18.095, -213.352,  73, 126}, -- Qufim Island (I-9)      {R}
}
local destination = nationDestinations[math.random(#randomDestinations)]
player:setPos(unpack(destination))
```
the nation enumerators are defined in scripts/globals/zone.lua

otherwise, nice clean code. thanks for the contribution!